nexus/src/app/instance.rs

```diff
@@ -1052,6 +1091,264 @@ impl super::Nexus {
+        // XDE devices for its NICs, and creating an XDE device for a virtual IP
+        // creates a V2P mapping that maps that IP to that sled. This is fine if
+        // migration succeeded, but if it failed, the instance is running on the
+        // source sled, and the incorrect mapping needs to be replaced.
+        //
+        // TODO(#3107): When XDE no longer creates mappings implicitly, this
+        // can be restricted to cases where an instance's sled has actually
+        // changed.
+        self.create_instance_v2p_mappings(
+            opctx,
+            instance_id,
+            new_runtime.sled_id,
+        )
+        .await?;
+
+        let (.., sled) = LookupPath::new(opctx, &self.db_datastore)
+            .sled_id(new_runtime.sled_id)
+            .fetch()
+            .await?;
+
+        self.instance_ensure_dpd_config(
+            opctx,
+            db_instance.id(),
+            &sled.address(),
+            None,
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Ensures that the Dendrite configuration for the supplied instance is
+    /// up-to-date.
+    ///
+    /// # Parameters
+    ///
+    /// - `opctx`: An operation context that grants read and list-children
+    ///   permissions on the identified instance.
+    /// - `instance_id`: The ID of the instance to act on.
+    /// - `sled_ip_address`: The internal IP address assigned to the sled's
+    ///   sled agent.
+    /// - `ip_index_filter`: An optional filter on the index into the instance's
+    ///   external IP array.
+    ///   - If this is `Some(n)`, this routine configures DPD state for only the
+    ///     Nth external IP in the collection returned from CRDB. The caller is
+    ///     responsible for ensuring that the IP collection has stable indices
+    ///     when making this call.
+    ///   - If this is `None`, this routine configures DPD for all external
+    ///     IPs.
+    pub(crate) async fn instance_ensure_dpd_config(
+        &self,
+        opctx: &OpContext,
+        instance_id: Uuid,
+        sled_ip_address: &std::net::SocketAddrV6,
+        ip_index_filter: Option<usize>,
+    ) -> Result<(), Error> {
+        let log = &self.log;
+        let dpd_client = &self.dpd_client;
+
+        info!(log, "looking up instance's primary network interface";
+              "instance_id" => %instance_id);
+
+        let (.., authz_instance) = LookupPath::new(opctx, &self.db_datastore)
+            .instance_id(instance_id)
+            .lookup_for(authz::Action::ListChildren)
+            .await?;
+
+        let network_interface = match self
+            .db_datastore
+            .derive_guest_network_interface_info(&opctx, &authz_instance)
+            .await?
+            .into_iter()
+            .find(|interface| interface.primary)
+        {
+            Some(interface) => interface,
+            // Return early if instance does not have a primary network
+            // interface
```

Comment:
why?

Reply:
I'm not sure--most of this part of the change is copied from `sic_instance_create`. My guess is that it's possible in principle to have external IPs assigned to an instance without actually having any NICs attached, and this is this routine's way of handling that case.

I tweaked the comment a bit in aa982b2d to reflect this understanding.

Reply:
:+1: